### PR TITLE
HPCC-16543 Improve error reporting of JlibSemTest and implement JlibSemTestStress.

### PR DESCRIPTION
**Pull request by AttilaVamos**
Add time and taken values to ASSERT log

Add semaphore stress test

Signed-off-by: Attila Vamos attila.vamos@gmail.com

### CODE REVIEW

**Comment by AttilaVamos** [comment]
@jakesmith please review.

File: testing/unittests/jlibtests.cpp [hunk 1]
```diff
@@ -57,12 +57,15 @@ class JlibSemTest : public CppUnit::TestFixture
...
         ASSERT(taken < 5);
     }
+
     void testTimedElapsed(Semaphore & sem, unsigned time)
     {
         unsigned now = msTick();
         sem.wait(time);
         unsigned taken = msTick() - now;
-        ASSERT(taken >= time && taken < 2*time);
+        VStringBuffer errMsg("values: time: %d, taken: %d", time, taken);
```
Comment (jakesmith) [inline testing/unittests/jlibtests.cpp]: trivial: %u more correct.

File: testing/unittests/jlibtests.cpp [hunk 2]
```diff
@@ -85,6 +88,63 @@ class JlibSemTest : public CppUnit::TestFixture
...
+
+protected:
+
+    void testSetup()
+    {
+    }
+
+    void testCleanup()
+    {
+    }
```
Comment (jakesmith) [inline testing/unittests/jlibtests.cpp]: Not new, but good time to clearup/remove empty test.

File: testing/unittests/jlibtests.cpp [hunk 3]
```diff
@@ -85,6 +88,63 @@ class JlibSemTest : public CppUnit::TestFixture
...
+        CPPUNIT_TEST(testSetup);
+        CPPUNIT_TEST(testSimple);
+        CPPUNIT_TEST(testCleanup);
+    CPPUNIT_TEST_SUITE_END();
+
+protected:
+
+    void testSetup()
+    {
+    }
```
Comment (jakesmith) [inline testing/unittests/jlibtests.cpp]: Not new, but good time to clearup/remove empty test.

File: testing/unittests/jlibtests.cpp [hunk 4]
```diff
@@ -85,6 +88,63 @@ class JlibSemTest : public CppUnit::TestFixture
...
+        unsigned long sumTaken = 0;
+        unsigned maxTaken = 0;
+        unsigned timeLimit = 2 * time;
+        unsigned numberOfOut = 0;
+
+        PROGLOG("Start loop");
+        for (int i = 0 ; i <= loopCount; i++)
+        {
+            unsigned now = msTick();
+            sem.wait(time);
```
Comment (jakesmith) [inline testing/unittests/jlibtests.cpp]: pedantic: could test if signalled, which would be unexpected/error.

File: testing/unittests/jlibtests.cpp [hunk 5]
```diff
@@ -85,6 +88,63 @@ class JlibSemTest : public CppUnit::TestFixture
...
+    {
+    }
+
+    void testCleanup()
+    {
+    }
+
+    void testTimedElapsed(Semaphore & sem, unsigned time, unsigned loopCount)
+    {
+        unsigned long sumTaken = 0;
```
Comment (jakesmith) [inline testing/unittests/jlibtests.cpp]: minor: if you want a 8 byte integer, we use a unsigned __int64, which guarantees it,

File: testing/unittests/jlibtests.cpp [hunk 6]
```diff
@@ -85,6 +88,63 @@ class JlibSemTest : public CppUnit::TestFixture
...
+        CPPUNIT_ASSERT_MESSAGE(errMsg.str(), 0 == numberOfOut);
+        PROGLOG("%s", errMsg.str());
+    }
+
+    void testSimple()
+    {
+        //Very basic semaphore stress tests.
+        Semaphore sem;
+        sem.signal();
+        sem.wait();
```
Comment (jakesmith) [inline testing/unittests/jlibtests.cpp]: minor: should test that actually signalled: (if (sem.wait())

**Comment by jakesmith** [comment]
@AttilaVamos - a few trivial/minor comments

File: testing/unittests/jlibtests.cpp (hunk 1 above)
Comment (AttilaVamos) [inline testing/unittests/jlibtests.cpp]: Ok

File: testing/unittests/jlibtests.cpp (hunk 3 above)
Comment (AttilaVamos) [inline testing/unittests/jlibtests.cpp]: Removed

File: testing/unittests/jlibtests.cpp (hunk 4 above)
Comment (AttilaVamos) [inline testing/unittests/jlibtests.cpp]: Implemented

File: testing/unittests/jlibtests.cpp (hunk 6 above)
Comment (AttilaVamos) [inline testing/unittests/jlibtests.cpp]: Implemented

**Comment by AttilaVamos** [comment]
@jakesmith updated, please check again.

File: testing/unittests/jlibtests.cpp [hunk 7]
```diff
@@ -93,42 +83,37 @@ class JlibSemTestStress : public CppUnit::TestFixture
...
             maxTaken = (taken > maxTaken ? taken : maxTaken);
             numberOfOut += (taken > timeLimit ? 1 : 0);
         }
 
-        VStringBuffer errMsg("values: time: %d, loop: %d, sumTaken: %ld, average taken: %ld, max taken: %d, out of limit: %d times",
-                                time, loopCount, sumTaken, sumTaken/loopCount, maxTaken, numberOfOut );
-        CPPUNIT_ASSERT_MESSAGE(errMsg.str(), 0 == numberOfOut);
+        VStringBuffer errMsg("values: time: %d, loop: %d, sum taken: %llu, average taken: %llu, max taken: %d, out of limit: %d times, signaled: %s",
+                                time, loopCount, sumTaken, sumTaken/loopCount, maxTaken, numberOfOut, (isSignaled ? "yes" : "no"));
+        CPPUNIT_ASSERT_MESSAGE(errMsg.str(), 0 == numberOfOut || isSignaled );
```
Comment (jakesmith) [inline testing/unittests/jlibtests.cpp]: This condition look wrong.
Should be : 0 == numberOfOut && !isSignaled

File: testing/unittests/jlibtests.cpp (hunk 6 above)
Comment (jakesmith) [inline testing/unittests/jlibtests.cpp]: Right, but probably should have timeout here e.g. if (!sem.wait(1000)) error, better than stalling.
Should never have to wait of course.

**Comment by jakesmith** [comment]
@AttilaVamos - one issue. If you want to go ahead and squash after changing that, then will be ready to merge.

**Comment by AttilaVamos** [comment]
@jakesmith updated and squashed, please take a look on it.

File: testing/unittests/jlibtests.cpp [hunk 8]
```diff
@@ -85,6 +78,62 @@ class JlibSemTest : public CppUnit::TestFixture
...
+
+    void testSimple()
+    {
+        //Very basic semaphore stress tests.
+        Semaphore sem;
+        sem.signal();
+        if (!sem.wait(1000))
+        {
+            VStringBuffer errMsg("Semaphore stalled (%s:%s)", __FILE__, __LINE__);
+            CPPUNIT_ASSERT_MESSAGE(errMsg.str(), 0);
```
Comment (jakesmith) [inline testing/unittests/jlibtests.cpp]: trivial: Could you have used CPPUNIT_FAIL, since you have no condition here?

**Review by jakesmith — APPROVED** [review]

**Comment by jakesmith** [comment]
@richardkchapman - good to merge.

**Comment by AttilaVamos** [comment]
@richardkchapman please wait with merge until I will change to CPPUNIT_FAIL().

**Comment by AttilaVamos** [comment]
@richardkchapman please merge.

**Pull request merged**